### PR TITLE
WIP: Add subgroups feature

### PR DESCRIPTION
**Pull request by alan-baker**
* API
  * new feature: subgroups
  * new properties for adapter info:
    * minSubgroupSize
    * maxSubgroupSize
* WGSL
  * new built-in values
    * subgroup_invocation_id
    * subgroup_size
  * subgroup and quad built-in functions
  * new uniformity diagnostic subgroup_uniformity

Missing subgroup_branching diagnostic and uniformity updates.

### CODE REVIEW

**Comment by alan-baker** [comment]
@kainino0x @toji I'd appreciate some early feedback on the API changes around properties.

File: spec/index.bs [hunk 1]
```diff
@@ -1929,6 +1931,29 @@ interface GPUAdapterInfo {
         this value is not recommended. Applications which change their behavior based on the
         {{GPUAdapterInfo}}, such as applying workarounds for known driver issues, should rely on the
         other fields when possible.
+
+    : <dfn>properties</dfn>
```
Comment (dneto0) [inline spec/index.bs:1935]: Is there a need/purpose for encapsulating these things in a new GPUAdapterProperties object?

Can the two new fields be included directly in GPUAdapterInfo?

Somewhat related in C-style APIs, I expect we would have a new chained struct, e.g. either GPUAdapterSubgroupInfo  (or GPUAdapterPropertiesSubgroupInfo if we do introduce a new GPUAdapterProperties)

File: wgsl/index.bs [hunk 2]
```diff
@@ -12172,6 +12263,48 @@ executing a [=statement/discard=] statement), execution of the quad may be
...
+Invocations that are not executing code due to control flow are considered
+<dfn noexport dfn-for="subgroups">inactive</dfn>.
+When the subgroup size is greater than the number of invocation in a subgroup, the extra invocations are
+considered inactive.
+[=Helper invocations=] may be active or inactive in an operation.
+That is, on some devices helper invocations may participate in subgroup operations,
+while on other devices they may not.
+
+Note: There is considerable non-portability among underlying devices when
+operating in non-uniform control flow and device compilers often agressively
```
Comment (dneto0) [inline wgsl/index.bs:12303]: "aggressively"

File: wgsl/index.bs [hunk 3]
```diff
@@ -12172,6 +12263,48 @@ executing a [=statement/discard=] statement), execution of the quad may be
...
+optimize such code.
+The result is that subgroups may contain different invocations than the shader author
+reasons ought to be active in a particular operation.
```
Comment (dneto0) [inline wgsl/index.bs:12306]: The "reasons ought to" phrasing seems awkward.

Suggest replacing this sentence with something like one of these:

> The result is that the set of active invocations executing a particular subgroup operation may not be what a shader author expects.

> The result is that for a particular subgroup operation, the set of active invocations may not be what a shader author expects.

> The result is that for a particular subgroup operation, the set of active invocations may differ from shader author expectations.

**Comment by alan-baker** [comment]
Closed in favour of #4963

**Review by dneto0 — COMMENTED** [review]
I see this is now closed .

Submitting reivew to save partial results.

**Pull request closed without merging**